File: src/guard.test.ts
// Unit tests for the guard's two load-bearing pure parts: signature verification (what "verified"
// actually means) and OIDC token discovery (the thing that decides whether a pipeline needs a
// stored secret at all).

import assert from "node:assert/strict";
import { test } from "node:test";
import { ed25519 } from "@noble/curves/ed25519";
import { getIdToken, looksLikeCi, OidcError } from "./oidc.js";
import { keyFingerprint, signedPayload, verifyBundle } from "./verify.js";

const hex = (b: Uint8Array) => Buffer.from(b).toString("hex");

function sign(version: number, policy: string) {
  const priv = ed25519.utils.randomPrivateKey();
  const pub = ed25519.getPublicKey(priv);
  const sig = ed25519.sign(signedPayload(version, policy), priv);
  return { bundle: { version, policy, sig: hex(sig) }, pubkey: hex(pub) };
}

// ---------------------------------------------------------------------------------------------
// Verification
// ---------------------------------------------------------------------------------------------

test("the signed payload is exactly `version\\npolicy` — the contract with the console", () => {
  // If this ever changes on one side only, the guard silently rejects every real policy (or worse,
  // accepts a forged one). Pinned here so a divergence is a failing test, not a production mystery.
  assert.equal(
    new TextDecoder().decode(signedPayload(42, "permit(principal, action, resource);")),
    "42\npermit(principal, action, resource);",
  );
});

test("accepts a real signature from the console's signer", () => {
  // Vector produced by @deter/server's signBundle with a fixed 32-byte key (all 0x42) — the same
  // one the broker's Rust verifier is pinned against. Three independent implementations, one
  // signature.
  const pubkey = "2152f8d19b791d24453242e15f2eab6cb7cffa7b6a5ed30097960e069881db12";
  const policy =
    'permit(principal == Deter::"sandbox", action == Action::"connect", resource == Net::"egress") when { context.host == "api.anthropic.com" };\n';
  const sig =
    "a5679c310de8a984445c45c5d3fd9cdc9b85068d8eace3cf99e08dd695c69723cf4c0237d964f0e17d94c1d7104d06f3b15d425c70a20beed2b4ebcf95125c0c";
  assert.deepEqual(verifyBundle({ version: 100, policy, sig }, pubkey), { ok: true });
});

test("rejects a tampered policy, version, or signature", () => {
  const { bundle, pubkey } = sign(7, 'permit(principal, action, resource) when { context.host == "ok.com" };');
  assert.ok(verifyBundle(bundle, pubkey).ok);

  // The version is INSIDE the signature, so a rollback attack changes the payload.
  assert.equal(verifyBundle({ ...bundle, version: 6 }, pubkey).ok, false);

  // Widening the policy under the original signature must fail — this is the whole point.
  const widened = bundle.policy.replace("ok.com", "evil.com");
  assert.equal(verifyBundle({ ...bundle, policy: widened }, pubkey).ok, false);

  // A different key doesn't verify.
  const other = sign(7, "x");
  assert.equal(verifyBundle(bundle, other.pubkey).ok, false);
});

test("malformed input is a verdict, never a throw", () => {
  const { bundle, pubkey } = sign(1, "p");
  for (const [b, k, why] of [
    [{ ...bundle, sig: "nothex" }, pubkey, "non-hex signature"],
    [{ ...bundle, sig: "abcd" }, pubkey, "short signature"],
    [bundle, "abcd", "short key"],
    [bundle, "zz".repeat(32), "non-hex key"],
    [{ version: 1, policy: "p" } as never, pubkey, "missing signature"],
    [{ version: "1", policy: "p", sig: bundle.sig } as never, pubkey, "non-integer version"],
  ] as const) {
    const v = verifyBundle(b, k);
    assert.equal(v.ok, false, `expected a failed verdict for ${why}`);
    assert.ok(v.ok === false && v.reason, `${why} should explain itself`);
  }
});

test("a key fingerprint is short and doesn't leak the whole key into logs", () => {
  const key = "a".repeat(32) + "b".repeat(32);
  const fp = keyFingerprint(key);
  assert.ok(fp.length < 20);
  assert.notEqual(fp, key);
  assert.equal(keyFingerprint("short"), "short");
});

// ---------------------------------------------------------------------------------------------
// Credential discovery
// ---------------------------------------------------------------------------------------------

test("an explicitly supplied ID token wins, so any platform can opt in", async () => {
  const t = await getIdToken("deter-console", { DETER_ID_TOKEN: "  tok  " } as NodeJS.ProcessEnv);
  assert.deepEqual(t, { token: "tok", source: "explicit" });
});

test("GitLab gets the exact snippet it needs, not a description of it", async () => {
  await assert.rejects(
    () => getIdToken("deter-console", { GITLAB_CI: "true" } as NodeJS.ProcessEnv),
    (e: Error) => {
      assert.ok(e instanceof OidcError);
      // The failure has to be actionable: GitLab can't be auto-detected because the job must name
      // the token, so the message carries the YAML.
      assert.match(e.message, /id_tokens:/);
      assert.match(e.message, /DETER_ID_TOKEN/);
      assert.match(e.message, /deter-console/);
      return true;
    },
  );
});

test("with no identity at all, the error names every way to provide one", async () => {
  await assert.rejects(
    () => getIdToken("deter-console", {} as NodeJS.ProcessEnv),
    (e: Error) => {
      assert.match(e.message, /id-token: write/);
      assert.match(e.message, /DETER_ID_TOKEN/);
      assert.match(e.message, /DETER_CI_TOKEN/);
      return true;
    },
  );
});

test("CI detection covers the platforms we tailor messages for", () => {
  assert.ok(looksLikeCi({ GITHUB_ACTIONS: "true" } as NodeJS.ProcessEnv));
  assert.ok(looksLikeCi({ GITLAB_CI: "true" } as NodeJS.ProcessEnv));
  assert.ok(looksLikeCi({ JENKINS_URL: "http://x" } as NodeJS.ProcessEnv));
  assert.ok(looksLikeCi({ CI: "1" } as NodeJS.ProcessEnv));
  assert.ok(!looksLikeCi({} as NodeJS.ProcessEnv));
});

File: guard_test.go
package main

import (
	"context"
	"crypto/ed25519"
	"encoding/hex"
	"strings"
	"testing"
)

// ---------------------------------------------------------------------------------------------
// Verification
// ---------------------------------------------------------------------------------------------

func TestSignedPayloadIsTheContract(t *testing.T) {
	// If this ever changes on one side only, the guard silently rejects every real policy — or worse,
	// accepts a forged one. Pinned so a divergence is a failing test, not a production mystery.
	got := string(signedPayload(42, "permit(principal, action, resource);"))
	want := "42\npermit(principal, action, resource);"
	if got != want {
		t.Fatalf("signed payload = %q, want %q", got, want)
	}
}

func TestAcceptsConsoleSignature(t *testing.T) {
	// Vector produced by the console's TypeScript signer (@noble/curves) with a fixed 32-byte key
	// (all 0x42) — the same vector the broker's Rust verifier is pinned against. Three
	// implementations, one signature; ed25519 is deterministic, so they must agree byte for byte.
	const pub = "2152f8d19b791d24453242e15f2eab6cb7cffa7b6a5ed30097960e069881db12"
	const policy = "permit(principal == Deter::\"sandbox\", action == Action::\"connect\", resource == Net::\"egress\") when { context.host == \"api.anthropic.com\" };\n"
	const sig = "a5679c310de8a984445c45c5d3fd9cdc9b85068d8eace3cf99e08dd695c69723cf4c0237d964f0e17d94c1d7104d06f3b15d425c70a20beed2b4ebcf95125c0c"

	if err := verifyBundle(SignedBundle{Version: 100, Policy: policy, Sig: sig}, pub); err != nil {
		t.Fatalf("a real console signature must verify, got: %v", err)
	}
}

func signFixture(t *testing.T, version int64, policy string) (SignedBundle, string) {
	t.Helper()
	pub, priv, err := ed25519.GenerateKey(nil)
	if err != nil {
		t.Fatal(err)
	}
	sig := ed25519.Sign(priv, signedPayload(version, policy))
	return SignedBundle{Version: version, Policy: policy, Sig: hex.EncodeToString(sig)},
		hex.EncodeToString(pub)
}

func TestRejectsTampering(t *testing.T) {
	b, pub := signFixture(t, 7, `permit(principal, action, resource) when { context.host == "ok.com" };`)
	if err := verifyBundle(b, pub); err != nil {
		t.Fatalf("baseline should verify: %v", err)
	}

	// The version is INSIDE the signature, so a rollback changes the payload.
	rolled := b
	rolled.Version = 6
	if err := verifyBundle(rolled, pub); err == nil {
		t.Error("a rolled-back version must not verify")
	}

	// Widening the policy under the original signature must fail — this is the whole point.
	widened := b
	widened.Policy = strings.ReplaceAll(b.Policy, "ok.com", "evil.com")
	if err := verifyBundle(widened, pub); err == nil {
		t.Error("a widened policy must not verify")
	}

	// A different key doesn't verify.
	_, other := signFixture(t, 7, "x")
	if err := verifyBundle(b, other); err == nil {
		t.Error("a foreign key must not verify")
	}
}

func TestMalformedInputIsAnErrorNotAPanic(t *testing.T) {
	b, pub := signFixture(t, 1, "p")
	cases := []struct {
		name   string
		bundle SignedBundle
		key    string
	}{
		{"non-hex signature", SignedBundle{1, "p", "nothex"}, pub},
		{"short signature", SignedBundle{1, "p", "abcd"}, pub},
		{"short key", b, "abcd"},
		{"non-hex key", b, strings.Repeat("zz", 32)},
		{"no signature", SignedBundle{Version: 1, Policy: "p"}, pub},
		{"no policy", SignedBundle{Version: 1, Sig: b.Sig}, pub},
	}
	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			if err := verifyBundle(c.bundle, c.key); err == nil {
				t.Fatal("expected an error")
			}
		})
	}
}

func TestKeyFingerprintDoesNotLeakTheKey(t *testing.T) {
	key := strings.Repeat("a", 32) + strings.Repeat("b", 32)
	fp := keyFingerprint(key)
	if fp == key {
		t.Error("fingerprint must not be the whole key")
	}
	if len([]rune(fp)) > 20 {
		t.Errorf("fingerprint too long for a log line: %q", fp)
	}
	if keyFingerprint("short") != "short" {
		t.Error("a short value should pass through")
	}
}

// ---------------------------------------------------------------------------------------------
// Credential discovery
// ---------------------------------------------------------------------------------------------

func envFrom(m map[string]string) func(string) string {
	return func(k string) string { return m[k] }
}

func TestExplicitTokenWins(t *testing.T) {
	// Any platform can opt in by setting this, which is what makes GitLab (and Buildkite, and
	// anything else) work without the guard knowing about them.
	got, err := getIDToken(context.Background(), defaultAudience,
		envFrom(map[string]string{"DETER_ID_TOKEN": "  tok  "}))
	if err != nil {
		t.Fatal(err)
	}
	if got.token != "tok" || got.source != srcExplicit {
		t.Fatalf("got %+v", got)
	}
}

func TestGitLabGetsTheExactSnippet(t *testing.T) {
	_, err := getIDToken(context.Background(), defaultAudience,
		envFrom(map[string]string{"GITLAB_CI": "true"}))
	if err == nil {
		t.Fatal("expected an error")
	}
	if !isOidcError(err) {
		t.Fatalf("expected an oidcError, got %T", err)
	}
	// GitLab can't be auto-detected because the job must name the token, so the message has to
	// carry the YAML rather than describe it.
	for _, want := range []string{"id_tokens:", "DETER_ID_TOKEN", defaultAudience} {
		if !strings.Contains(err.Error(), want) {
			t.Errorf("message should mention %q, got:\n%s", want, err)
		}
	}
}

func TestNoIdentityNamesEveryOption(t *testing.T) {
	_, err := getIDToken(context.Background(), defaultAudience, envFrom(nil))
	if err == nil {
		t.Fatal("expected an error")
	}
	for _, want := range []string{"id-token: write", "DETER_ID_TOKEN", "DETER_CI_TOKEN"} {
		if !strings.Contains(err.Error(), want) {
			t.Errorf("message should mention %q, got:\n%s", want, err)
		}
	}
}

func TestLooksLikeCI(t *testing.T) {
	for _, k := range []string{"CI", "GITHUB_ACTIONS", "GITLAB_CI", "JENKINS_URL"} {
		if !looksLikeCI(envFrom(map[string]string{k: "1"})) {
			t.Errorf("%s should be detected", k)
		}
	}
	if looksLikeCI(envFrom(nil)) {
		t.Error("an empty environment is not CI")
	}
}

// ---------------------------------------------------------------------------------------------
// Wire shapes
// ---------------------------------------------------------------------------------------------

func TestBaseURLTrimsTrailingSlashes(t *testing.T) {
	// A pasted console URL with a trailing slash would otherwise produce `//api/ci/policy`, which
	// some proxies treat as a different path.
	for in, want := range map[string]string{
		"https://c.example.com/":   "https://c.example.com",
		"https://c.example.com///": "https://c.example.com",
		"  https://c.example.com ": "https://c.example.com",
		"https://c.example.com":    "https://c.example.com",
	} {
		if got := baseURL(in); got != want {
			t.Errorf("baseURL(%q) = %q, want %q", in, got, want)
		}
	}
}
